batch_builder.go

```diff
@@ -0,0 +1,228 @@
+package hippo
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"sort"
+)
+
+// BatchBuilder is responsible for building a batch that will serialize within a desired size if possible.
+// It'll fit big and small populators together, but doesn't try to do so optimally.
+type BatchBuilder struct {
+	desiredSize       int
+	buf               *bytes.Buffer
+	serializedUpserts [][]byte
+	batchGUID         string
+	replaceAll        bool
+	ttlMinutes        uint32
+	builtBatchesCount int
+	hasClosedBatch    bool
+	sender            TagBatchPartSender // optional - to help track batch origin
+}
+
+// NewBatchBuilder builds a new BatchBuilder
+func NewBatchBuilder(desiredSize int, replaceAll bool, ttlMinutes uint32) *BatchBuilder {
+	return &BatchBuilder{
+		desiredSize:       desiredSize,
+		serializedUpserts: make([][]byte, 0),
+		replaceAll:        replaceAll,
+		ttlMinutes:        ttlMinutes,
+		buf:               bytes.NewBuffer(make([]byte, 0, desiredSize)),
+	}
+}
+
+// Reset resets the batch builder to build a new batch, reusing the underlying buffer.
+func (b *BatchBuilder) Reset(desiredSize int, replaceAll bool, ttlMinutes uint32) {
+	b.desiredSize = desiredSize
+	b.serializedUpserts = b.serializedUpserts[:0]
+	b.batchGUID = ""
+	b.replaceAll = replaceAll
+	b.ttlMinutes = ttlMinutes
+	b.builtBatchesCount = 0
+	b.hasClosedBatch = false
+}
+
+// SetSenderInfo sets optional metadata about the service sending batches
+func (b *BatchBuilder) SetSenderInfo(sender TagBatchPartSender) {
+	b.sender = sender
+}
+
+func (b *BatchBuilder) isSenderInfoSet() bool {
+	return b.sender.ServiceName != "" || b.sender.ServiceInstance != "" || b.sender.HostName != ""
+}
+
+// AddUpsert attempts to add the input upsert into the batch.
+// Make sure to add all upserts to the batch before calling BuildBatch()
+func (b *BatchBuilder) AddUpsert(upsert *TagUpsert) error {
+	ser, err := json.Marshal(upsert)
+	if err != nil {
+		return fmt.Errorf("Error serializing TagUpsert: %s", err)
+	}
+	b.serializedUpserts = append(b.serializedUpserts, ser)
+	return nil
+}
+
+// SetBatchGUID sets the GUID that was returned after submitting the first part to the server.
+func (b *BatchBuilder) SetBatchGUID(guid string) {
+	b.batchGUID = guid
+}
+
+// BuildBatch builds and returns a serialized batch.
```

Comment:
The terminology of this felt a little bit confusing to me as I was reading through most of this... I am familiar with this overall goal of this struct, but reading it from an outside perspective might be a bit ambiguous towards what a "batch" is in the different context.

In this specific receiver function, this actually seems to build a serialize _portion_ of the "overall" batch to be sent? Since it doesn't build the entire batch in one call and requires multiple calls to do that.

Reply:
Ah, it's a batch part, not a batch. Fair. I'll update that verbiage later, thanks

Reply:
Updated:
https://github.com/kentik/gohippo/blob/38ed3bf4a4a42d4ce9a67b484791b8f51a1a45d8/batch_builder.go#L75-L78